### ISSUE TITLE
Subpixel_contours function does not fail when z_value is not in array and no output_path is provided

### ISSUE THREAD

**Issue by npucino** [GeoscienceAustralia/dea-notebooks#590]
Currently, when we provide a `z_value `that is not in the array, the `subpixel_contours `function will fail raising a `ValueError `only if the `outputh_path `is provided. 
This is because the error is in Fiona trying to write a geojson of an empty array.
If we do not provide the output_path parameter, thus, we do not write out a geojson, the function will not raise a `ValueError `but print a message saying ""Failed to generate contours: 0".

**It is not a big issue, but in case we run this function in for or while loops for a lot of scenarios or datasets, it would be useful to make it fail in both cases, so that we can manage its behaviour with Try/Except clauses.**

Here is what happens.
Have a look at the highlighted image, obviously "hazy", but still with discernible coastline under the cloud cover.

<img width="875" alt="rgbs" src="https://user-images.githubusercontent.com/44853536/84842383-50735d00-b088-11ea-8f8d-07e1a8eb34e5.PNG">

I compute band indices, here shown is AWEI_ns:
<img width="947" alt="awei_shs" src="https://user-images.githubusercontent.com/44853536/84842529-bb249880-b088-11ea-8728-bc94a605f0ce.PNG">

And here, the AWEI_ns distribution, where the **ZERO** value is represented by the green vertical line.
<img width="540" alt="awei_ns" src="https://user-images.githubusercontent.com/44853536/84842616-ee672780-b088-11ea-9b03-1b4af25d4f14.PNG">

When running the function:
```
contours_gdf_tmp = subpixel_contours(da=ds_tidal.isel(time=1)["AWEI_ns"],
                                 z_values=0,
                                 crs=ds_tidal.crs,
                                 output_path=savetxt,              
                                 affine=ds_tidal.geobox.transform,
                                 min_vertices=200
                                 )
```
This error comes up:

```
Operating in multiple z-value, single array mode
Failed to generate contours: 0
Writing contours to /home/jovyan/dev2/dea-notebooks/my_data/cow_2019-10-23_WI_otsu.geojson
/env/lib/python3.6/site-packages/pyproj/crs/crs.py:53: FutureWarning: '+init=<authority>:<code>' syntax is deprecated. '<authority>:<code>' is the preferred initialization method. When making the change, be mindful of axis order changes: https://pyproj4.github.io/pyproj/stable/gotchas.html#axis-order-changes-in-proj-6
  return _prepare_from_string(" ".join(pjargs))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-201-2f6deadd530e> in <module>
      4                                  output_path=savetxt,
      5                                  affine=ds_tidal.geobox.transform,
----> 6                                  min_vertices=200
      7                                  )

~/dev2/dea-notebooks/Scripts/dea_spatialtools.py in subpixel_contours(da, z_values, crs, affine, attribute_df, output_path, min_vertices, dim)
    505         print(f'Writing contours to {output_path}')
    506         contours_gdf.to_crs({'init': 'EPSG:4326'}).to_file(filename=output_path, 
--> 507                                                            driver='GeoJSON')
    508     if output_path and output_path.endswith('.shp'):
    509         print(f'Writing contours to {output_path}')

/env/lib/python3.6/site-packages/geopandas/geodataframe.py in to_file(self, filename, driver, schema, index, **kwargs)
    531         from geopandas.io.file import to_file
    532 
--> 533         to_file(self, filename, driver, schema, index, **kwargs)
    534 
    535     def to_crs(self, crs=None, epsg=None, inplace=False):

/env/lib/python3.6/site-packages/geopandas/io/file.py in to_file(df, filename, driver, schema, index, **kwargs)
    177         df = df.reset_index(drop=False)
    178     if schema is None:
--> 179         schema = infer_schema(df)
    180     with fiona_env():
    181         crs_wkt = None

/env/lib/python3.6/site-packages/geopandas/io/file.py in infer_schema(df)
    229 
    230     if df.empty:
--> 231         raise ValueError("Cannot write empty DataFrame to file.")
    232 
    233     # Since https://github.com/Toblerity/Fiona/issues/446 resolution,

ValueError: Cannot write empty DataFrame to file.

```
However, without the **output_path** parameter:

```
contours_gdf_tmp = subpixel_contours(da=ds_tidal.isel(time=1)["AWEI_ns"],
                                 z_values=0,
                                 crs=ds_tidal.crs,
                                     
                                 affine=ds_tidal.geobox.transform,
                                 min_vertices=200
                                 )
```
No ValueError is raised, and a message is printed:
```
Operating in multiple z-value, single array mode
Failed to generate contours: 0
```
I think would be good to make it fail even in the latter case.

Cheers!
Nick

**Comment by robbibt** [comment]
Thanks for the issue @npucino: I think the function needs an error to be raised earlier (before the data is written to file) on if no output contours are produced. That way it will catch the problem regardless of whether users want to export a file.

I'll put in a fix today or tomorrow!

**Comment by robbibt** [comment]
Hey @npucino, this is proving a tiny bit more complex than I'd originally expected: the function has a 'single z-value, multiple arrays' mode that automatically extracts of contours from every time-step in your dataset like this (i.e. without the `.isel(time=1)`):

```
contours_gdf_tmp = subpixel_contours(da=ds_tidal["AWEI_ns"],
                                 z_values=0,
                                 crs=ds_tidal.crs,
                                 output_path=savetxt,              
                                 affine=ds_tidal.geobox.transform,
                                 min_vertices=200
                                 )
```

This will automatically apply the function to every array along the "time" axis, and print out any timesteps that failed. This was designed for exactly the kind of scenario you're encountering: to ignore some time-steps that have no valid contours, and continue to extract contours for the remaining time-steps.

Similarly, the other 'single array, multiple z-values' mode will iterate through a list of z-values, and print out any z-value that failed. Because the advantage of these modes is being able to apply the function to a large set of either heterogeneous time-steps, or multiple z-values that may or may not be in the dataset, it's not ideal to raise an exception in the middle of this process otherwise the function will stop prematurely and not process any of the other results that users are interested in.

That said, it's still definitely buggy that an exception is raised only when trying to write the output to disk. I can see a few ways of dealing with this:

1. Raise an exception if a contour extraction fails only if a single timestep and z-value is being provided (like in your scenario)
2. Raise an exception only if _all_ timesteps/z-values produced no valid contours. This would fail in your scenario, but not fail if the user supplies a set of timesteps/z-values.
3. Have a parameter that allows the user to specify if they want to skip invalid timesteps/z-values with just a print statement (like the current approach), or raise an exception instead (so kind of like the `errors{‘ignore’, ‘raise’, ‘coerce’}` in Panda's `.to_datetime()`

**Comment by robbibt** [comment]
@npucino I went for both option 2 and option 3, as I think this combination is the most flexible. For your application it should raise an exception by default, but you can also specify `errors='raise'` to always raise an exception if any single contour fails!

PR here, I've sent you an invite to the repo in case you want to review/accept the PR: https://github.com/GeoscienceAustralia/dea-notebooks/pull/595

You should be able to get the latest version of the code by doing:
```
git checkout develop
git pull
git checkout --track origin/contour_fix
```

**Comment by robbibt** [comment]
Closed by #595

**Comment by npucino** [comment]
Thx @robbibt, adding the parameter was a good idea, so that people can choose to use the function business as usual or managing exceptions if needed. Cheers!
